### ISSUE TITLE
Stock reduced when 'Pending Payment' order is set to 'Cancelled'

### ISSUE THREAD

**Issue by jrick1229** [woocommerce/woocommerce#28505]
**Prerequisites (mark completed items with an [x]):**
- [x] I have have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**

When an order that is currently `Pending Payment` is set to `Cancelled`, stock is being reduced.

**Expected behavior**

Since `Pending Payment` does not decrease stock, cancelling that order should not touch stock at all.

**Actual behavior**

If an order is created during checkout that results in a `Pending Payment` status, the order does not decrease the stock levels of the product it contains. Once we change that status from `Pending Payment` to `Cancelled`, the stock is being decreased equal to the amount of items the order contains.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**

Steps to reproduce the bug:
1. Set a product's stock level to **5**
2. Enable PayPal Standard
2. Add product to the cart
3. Click 'Proceed to PayPal' to navigate to PayPal
4. X out of the PayPal window before navigating back to your site
5. Navigate to the new `Pending Payment` order
6. Change the status to `Cancelled`
7. See the order note stating that the stock for that product was reduced

**Screenshots**

`Pending Payment` -> `Cancelled` order (incorrect): https://d.pr/i/zU8MYc

`Processing` -> `Cancelled` order (correct): https://d.pr/i/bAZYOf

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**

<details>

```
WC Database Version: 4.7.1
WooCommerce: by Automattic – 4.7.1

### Theme ###

Name: Storefront
Version: 2.9.0
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔
```
</details>

**Comment by damiannep** [comment]
Another case of this in 3522699-zen

**Comment by liviujs** [comment]
CONFIRM there's a bug that Reduces stock when orders status from Pending Payment to Cancelled is set.
->WC 4.7.1
->Payment gateway: VivaWallet.

**Comment by atousa-najafi** [comment]
I have the same issue on my site

**Comment by senff** [comment]
Also reported in #26221869-HC

**Comment by alpha-pix** [comment]
Can confirm I have seen this with Global Payments HPP on manual cancellations of Pending orders. It does not occur when the order is allowed to cancel automatically after the time set in "Hold Stock (minutes)". So a workaround for some of this is to let them time out and maybe reduce the minutes.

This behaviour does not seem to happen with manually created orders. Instead another anomaly occurs.  
(a) When creating a new manual order with status Pending, the stock is REDUCED. This is potentially useful behaviour, but seems to contradict the documentation and the "Hold Stock (minutes)" setting.  
(b) When manually cancelling that manual Pending order, stock is not touched. If it is decided that the proper behaviour is that manual orders should behave differently to gateway-created orders, and have stock reduced on creation of a manual pending order, then on cancellation, they should correspondingly be INCREASED to revert.

**Comment by CHadetz** [comment]
I'm fairly certain this is the cause, introduced in 4.7.0:
`* Fix - Adjust stock even if `reduce_stock` meta is not set in `wc_maybe_reduce_stock_levels`. #27763`

**Comment by rwaynick** [comment]
Another report in 26222717-hc.

**Comment by juliaamosova** [comment]
Hi @jrick1229,

Thank you for taking the time to report this bug, we really appreciate your help. I can reproduce it on my end. 

I am using WooCommerce 4.8 and set up PayPal Standard to test this. 

Updating order status from `Pending payment` to `Cancelled` does reduce the stock indeed which is an incorrect behavior:

![paypal](https://user-images.githubusercontent.com/19143190/101995431-c6532c80-3c97-11eb-9811-ae2cac133e38.jpg)

According to @liviujs:

>Double checked same bug happens to ALL CANCELED orders not depending on payment gateway.
The bug is reproduced from states:
Processing->Cancelled
On Hold->Cancelled
Pending Payment->Cancelled

I could not confirm the above. On my end, the issue only happened when changing status from `Pending payment` to `Cancelled`. 

Some examples I've tried to try to reproduce:

<hr>

Updating order status from `Processing` to `Cancelled` using `Cash on delivery` updates the stock correctly after the order had been cancelled - the stock gets increased as expected:

![cod](https://user-images.githubusercontent.com/19143190/101995505-46799200-3c98-11eb-8c88-f2ae475ff7d2.jpg)

<hr>

Updating order status from `On hold` to `Cancelled` using `Bank transfer` updates the stock correctly after the order had been cancelled - the stock gets increased as expected:

![bacs](https://user-images.githubusercontent.com/19143190/101995507-4bd6dc80-3c98-11eb-98be-0a231dbd1034.jpg)

<hr>

In regards to stock not being updated correctly when switching order status from `Pending payment` to `Cancelled`, we will try our best to fix it as soon as possible.

**Comment by robertocyrino** [comment]
+1 report in 3588499-zen.

**Comment by vedanshujain** [comment]
This is fixed via #28620 and will be available with WC 4.9
